### PR TITLE
Feature/gy ccr retry and rate limit

### PR DESCRIPTION
**Pull request by fholzhauser**
This PR includes initial versions of Gy/Ro CCR retry and rate limiting features and the corresponding tests.

Description of the implementation are in CCR_retry_and_rate_limit.md

### CODE REVIEW

File: src/ergw_aaa_ro.erl [hunk 1]
```diff
@@ -146,8 +146,34 @@ invoke(_Service, {_, 'CCR-Terminate'}, Session0, Events, Opts) ->
...
-    diameter:call(Function, ?APP, Request, []).
+call(Request, #{rate_limit_jobs_queue_name := RateLimitQueue} = Config) ->
+    try jobs:run(RateLimitQueue, 
+        fun() -> call(Request, maps:remove(rate_limit_jobs_queue_name, Config)) end)
+    catch
+        error:timeout -> {error, rate_limit}
+    end;
+
+call(Request, #{max_retries := MaxRetries} = Config) when MaxRetries > 0 ->
+    call_with_retry(Request, Config, MaxRetries+1, diameter_session:sequence(), []);
```
[the remark this answers is not in the document]
Comment (fholzhauser) [reply src/ergw_aaa_ro.erl]: I don't really like this one either, but unfortunately, according to the traces I made, the handle_error callback (where the timeout is handled) does not have the diameter header. I haven't (yet) found another way to preserve the same end-to-end ID across the requests, than the ugly way (to pre-allocate it when retry can happen). For that, this seems to be the most proper (although undocumented) function. I'll dig further into it, and if I can find a way to get the header into that callback, then I'll update this part.

File: CCR_retry_and_rate_limit.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+Rate limiting
+=============
+Rate limiting is implemented in the ergw_aaa_ro module and can be applied per procedure. 
+
+Behaviour
+---------
+Rate limiting is depending on the jobs application. It requires a jobs queue configured and available at the time of the processing a request for which there is retry configured. This is normally done by configuring the jobs application in the release environment. The name of the configured queue is used in the *rate_limit_jobs_queue_name* configuration of the procedure. The the jobs queue should be configured with a rate regulator with the desired maximum rate, and a max time which limits how long a request can wait unti a "slot" to be sent. If this max time is exceeded, the request fails with **rate_limit** error. 
+
```
[the remark this answers is not in the document]
Comment (fholzhauser) [reply CCR_retry_and_rate_limit.md]: Sure, this is rather verbose indeed. Personally I prefer being as specific as possible against having something short, unless of course I have to type it many times, which is probably not the case here. But I'm fine with a short(er) version too, if we can come up with a good one.

[the remark this answers is not in the document]
Comment (mgumz) [reply CCR_retry_and_rate_limit.md]: well … there are only 2 hard problems in CS :)

i prefer short … if the module/package is sufficiently specific. i don't need `long_module_name_with_things:and_then_an_even_longer_function_or_other_name`.

[the remark this answers is not in the document]
Comment (mgumz) [reply CCR_retry_and_rate_limit.md]: after looking at the usage: `rate_limit_jobs_queue_name` is in the context of the `rate_limit, `jobs` is an implementation detail which does not really belong in there (lets say, tomorrow i ll write `chains` - my `jobs-like` framework to achieve the same, but better).

imho, `rate_limit_queue` is sufficient to designate the name of the `queue`

[the remark this answers is not in the document]
Comment (fholzhauser) [reply CCR_retry_and_rate_limit.md]: ```rate_limit_queue``` it is, thanks !

File: src/ergw_aaa_ro.erl [hunk 3]
```diff
@@ -188,15 +222,45 @@ prepare_request(Packet, _SvcName, {PeerRef, _}) ->
...
+		   origin_realm = {OR, _},
+		   origin_state_id = {OSid, _}} = Caps,
+
+    RetryCCRHdr = Header#diameter_header{
+        is_retransmitted = PeersTried /= [],
+        is_proxiable = true,
+        cmd_code = 272,
+        application_id = 4,
+        end_to_end_id = E2EId
+    },
```
[the remark this answers is not in the document]
Comment (fholzhauser) [reply src/ergw_aaa_ro.erl]: You are correct, this is a leftover from an initial (not working) approach. In this return format it is not needed. Fixed.

**Comment by RoadRunnr** [comment]
I think we all have made a mistake with the transmit to server logic. With DIAMTER it is quite easy to mix up peers and servers, and we have done that here....

Currently, we take the list of available DIAMETER peers as possible re-transmission targets. BUT a peer does not need to be a server (or the OCS). It could a routing agent or a proxy.

3GPP and RFC 4006 clearly state that the request should be send to a configure OCS. This means we have to set the `Destination-Host` AVP to the configured value and possibly the `Destination-Realm`.
Once a fail-over occurs, we switch to a backup server (not to a backup peer) by setting the `Destination-Host` AVP to a new value, and then let the peer selection and routing forward the request.

So, the current logic needs to be changed:
* change the `Destination-Host` config setting to support more than one server (we need to decide if this should be a list, a primary/backup tuple, or something else)
* use the list of configured Destination-Hosts in the fail-over handling (maybe with with priorities or load balancing)

Additional, it is not clear to me reading from the RFCs with we need to keep the End-to-End identifier. As I understand the RFCs the E2E id is only kept when re-sending to the same server. Since we actively select a new server, using a new E2E id seems to be required.
RFC 4006 even mentions that duplicate detection is supposed to use Session-Id AVP, and CC-Request-Number AVP (not the E2E id).

Comments?

**Comment by fholzhauser** [comment]
@RoadRunnr I think both retry target selection methods are valid, depending on the particular deployment. I'd propose to add the diameter host (instead of peer) selection as you described as an another retry option.

**Pull request merged**